### PR TITLE
Added FIFO ibyte support and code cleanups

### PR DESCRIPTION
**Pull request by stefanlinden**
This PR adds support for ibyte input to the FIFO signal_source block. The output is (still) converted to gr_complex.

Furthermore, an additional buffer (based on the one used in `gr_complex_ip_packet_source.cc`) improves performance of the FIFO pipe reading slightly. Do note that it could add latency to low sample rate configurations so this might need to be made configurable.

Finally, some code cleanup such as removing function definitions from the `fifo_reader` header file and applying `clang-format` and `clang-tidy`.

It is still not as fast as a file signal_source block and still some code is repeated between the 'interleaved' sample handling and the `gr_complex` sample handling, so future optimisations are definitely possible.

### CODE REVIEW

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h
Comment (jwmelto) [inline src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h:57]: Template definitions need to be visible at the point of instantiation. Most of the time, that means defining them in the header. It's possible that there is no harm, since this is a private method, but knowing that most developers aren't aware of these nuances, I prefer to comment at a minimum why it's safe to define this out of line.

An alternative to in-line definition is to move the definition after the class declaration in this file. That keeps the declaration easier to read, but satisfies the visibility requirements.

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h
Comment (jwmelto) [inline src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h:70]: don't use naked pointers. If you want dynamic sizing, use `std::vector<char>`. If you are ok with static sizing, use `std::array<char>`. Or, if you must, use `std::unique_ptr<char[]>` (see [SO](https://softwareengineering.stackexchange.com/questions/380176/should-i-use-a-unique-ptr-with-an-array-type-or-a-vector))

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h
Comment (jwmelto) [inline src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h:71]: please declare each variable in a separate statement. Note that `buffer_size_` is useless information, if you use a container

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h
Comment (jwmelto) [reply src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h:70]: or maybe you want a `boost::circular_buffer`?

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc
Comment (jwmelto) [inline src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc:37]: Just as a rule of thumb, if you are calling `new` explicitly, you've missed the modern C++ way. Either containers or `make_unique<T>`/`make_shared<T>` should be preferred.

Using RAII constructs has the (intentional) benefit of relieving you of destructor responsibilities.

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc
Comment (jwmelto) [inline src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc:64]: that's not a bad comment. Why remove it?

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc
Comment (jwmelto) [inline src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc:103]: I realize this is moved code, so your responsibility is limited. However, I'd make a point of noting that this will block until FIFO_SIZE is read completely. At least, I think that's the design.

Now, the question is, should you read more than `noutput_items`, since that's all you wanted?

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc
Comment (jwmelto) [inline src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc:124]: these copies can be replace with pointers:
```c++
auto real = reinterpret_cast<Type const*>(&fifo_buffer_[buffer_idx_]);
auto imag = reinterpret_cast<Type const*>(&fifo_buffer_[buffer_idx_ + sizeof(Type)]);
static_cast<gr_complex*>(output_items[0])[n] = gr_complex(*real, *imag);
```

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc
Comment (jwmelto) [inline src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc:188]: prefer RAII to explicit memory management.

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc
Comment (jwmelto) [reply src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc:103]: I do see the previous implementation only read `noutput_items` from the FIFO

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc
Comment (jwmelto) [inline src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc:148]: Note that the FIFO is buffered (`ifstream`) so you aren't gaining anything by buffering yourself here. If you have empirical evidence that reading one item at a time is "too slow" and that this buffering is much faster, then you owe future readers a comment to that effect. However, I'm hard pressed to see how this buffering -> single item extraction can be faster than the standard library.

**Review by jwmelto — COMMENTED** [review]
It seems like this change was much more intrusive than simply supporting the `ibyte` data type. Perhaps break it into two PRs, one for `ibyte` and another for performance changes (if they are needed)?

Stylistic differences aside, my own experience with data sources was a difficult learning curve. One of the main points (that isn't sufficiently documented in GNU Radio) is that a source should block until it can produce the requested number of elements. It can return earlier, if desired, but isn't expected to. The code here blocks until an arbitrary number of bytes are read, which could have unpleasant effects downstream, depending on the speed of the FIFO.

**Comment by stefanlinden** [comment]
Hi @jwmelto, thanks for the detailed code review, it is appreciated. As you can see my own background is mostly in c so a lot of the c++ nuances are new to me.
I've been doing some performance analyses between the 'new' version with buffering and the one 'old' implementation without buffering and indeed there is no significant difference based on GNSS-SDR's own timer.
Therefore, for simplicity's sake I will remove the buffer and apply the proposed changes that are still relevant.

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc
Comment (stefanlinden) [reply src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc:64]: Put back in [77c682e](https://github.com/gnss-sdr/gnss-sdr/pull/584/commits/77c682e8b312c645d77e60289b0d67b33e67c633).

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc
Comment (stefanlinden) [reply src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc]: Obsolete after [77c682e](https://github.com/gnss-sdr/gnss-sdr/pull/584/commits/77c682e8b312c645d77e60289b0d67b33e67c633).

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h
Comment (stefanlinden) [reply src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h]: Not applicable anymore after [77c682e](https://github.com/gnss-sdr/gnss-sdr/pull/584/commits/77c682e8b312c645d77e60289b0d67b33e67c633).

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h
Comment (stefanlinden) [reply src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h]: Not applicable anymore after [77c682e](https://github.com/gnss-sdr/gnss-sdr/pull/584/commits/77c682e8b312c645d77e60289b0d67b33e67c633).

File: src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h
Comment (stefanlinden) [reply src/algorithms/signal_source/gnuradio_blocks/fifo_reader.h:57]: Thanks. For observers, also see: https://stackoverflow.com/questions/495021/why-can-templates-only-be-implemented-in-the-header-file.

**Comment by carlesfernandez** [comment]
Thanks for your work, the PR seems fine to me. Let's wait for @jwmelto  approval.

That's great because we want to make a new release in the following days, so all the changes in the current `next` branch get into the forthcoming GNU/Linux distro releases. I hope to get this PR in.

**Review by jwmelto — APPROVED** [review]
Looks great.  Thanks.

**Pull request merged**
